controllers/machinedeletionremediation_controller.go

```diff
@@ -282,37 +281,43 @@ func (r *MachineDeletionRemediationReconciler) getNodeFromMdr(ctx context.Contex
 	return node, nil
 }
 
-// getMachine retrieves a machine from the cluster based on the remediation.
-// It returns the machine, a boolean indicating whether the machine was expected to be found,
-// and an error if any occurred during the retrieval process.
+// getMachine retrieves a Machine from the cluster based on the remediation.
+// It returns the machine and an error if any occurred during the retrieval process.
 func (r *MachineDeletionRemediationReconciler) getMachine(ctx context.Context, remediation *v1alpha1.MachineDeletionRemediation) (*machinev1beta1.Machine, error) {
-	// If the Machine's Name and Ns come from the related Node, it is expected
-	// to find the Machine in the cluster, while if its Name and Ns come from
-	// CR's annotation, the Machine might have been deleted upon our request.
+	// The Name and Namespace to retrieve the target Machine can come from the following sources:
+	// - the remediation's ownerReference: if the remediation was created by MachineHealthcheck
+	// - the remediation's Node: if the remediation was created by NodeHealthcheck or manually
+	// - the remediation's MachineNameNsAnnotation annotation: once the Machine is found and its Name and Namespace are saved in
+
+	// Try to get first the Machine's data from the remediation's annotation, if any. It means that the Machine was already
+	// been found in a previous cycle and we can use the data to verify if the Machine was deleted upon our request or not
 	machineName, machineNs, err := getRemediationDataFromAnnotation(remediation, MachineNameNsAnnotation)
 	if err != nil {
 		r.Log.Error(err, "could not get Machine data from remediation", "remediation", remediation.GetName(), "annotation", MachineNameNsAnnotation)
 		return nil, unrecoverableError
 	}
 
-	var mustExist bool
-	if machineName == "" {
-		// Remediation does not have the MachineDataAnnotation yet. We will try to get the Machine's data from its Node,
-		// and we expect to find the Node and the Machine in the cluster.
-		mustExist = true
+	// If the Machine's Name is not in the annotation, it means that it must come from the other two sources and in
+	// turns it means that the Machine must exist in the cluster, otherwise an error is returned.
+	mustExist := machineName == ""
+
+	if machineName == "" && len(remediation.GetOwnerReferences()) > 0 {
+		machineName, machineNs = r.getMachineNameNsFromOwnerReference(ctx, remediation)
+		if machineName != "" {
+			r.Log.Info("MHC created CR: target machine is CR ownerReference", "machine", machineName, "namespace", machineNs)
+		}
+	}
 
-		node, err := r.getNodeFromMdr(ctx, remediation)
+	if machineName == "" {
```

Comment:
I think this pattern is very confusing. 
i.e: 

```
if x == something {
//Do some code
}
//Assuming x value may have changed in the previous code block
if x == something {
}
```

Reply:
maybe something like:
```
machineNameFromAnnotation, machineNs, err := getRemediationDataFromAnnotation(remediation, MachineNameNsAnnotation)
...
if machineNameFromAnnotation == "" {
  if machineNameFromOR, machineNs = r.getMachineNameNsFromOwnerReference(ctx, remediation); machineNameFromOR != "" {
    r.Log.Info("MHC created CR: target machine is CR ownerReference", "machine", machineNameFromOR, "namespace", machineNs)
} else {
    r.Log.Info("trying to get target Machine from the Node", "node", remediation.Name)
    machineNameFromCR, machineNs, err = r.getMachineNameNsFromCR(ctx, remediation)
    ...
}
```

Reply:
The problem here is that later I need `machineName` and `machineNs` variables to be valorized to get the Machine.
However, I understand the problem and will try to figure out something clearer

Reply:
Maybe put all 3 options to get the name/ns into its own functions?
And have 1 wrapper function?

`func getMachineNamespacedName() (*NamespacedName, bool, error)`
3 x `func getMachineNamespacedNameFrom...() (*NamespacedName, error)`


```
func getMachine() ... {
  var mustExist bool
  var machineNsName *NamespacedName
  if machineNsName, err := getMachineNamespacedName(); err != nil {
    return err
  }
  // go on mustExist and machineNsName
  ...
}

func getMachineNamespacedName(...)  (*NamespacedName, bool, error) {
  mustExist = false
  if machineNsName, err = GetMachineNamespacedNameFrom...(...); err != nil {
    return nil, mustExist, err
  } else if machineNsName != nil {
    return machineNsName, mustExist, nil
  }
  mustExist = true
  // repeat for other 2...
}

func GetMachineNamespacedNameFrom...(...) (*NamespacedName, error) {
  ...
}
```

Can also very well be that this is overengineered 🙂

Reply:
does [16570a4](https://github.com/medik8s/machine-deletion-remediation/pull/111/commits/16570a4ab22f7b8232f9a2d127c6b3ec9ccda323) looks better?

Reply:
I have no strong preference here, most important are the good comments 👍🏼